### PR TITLE
Add status key for report test pass/failure based on returncode or regex

### PR DESCRIPTION
**Pull request by shahzebsiddiqui**
This PR is not complete but i wanted to share my starting point and was hoping i can get some assistance in a bug i am getting.

This PR tries to clean up some of the code base by doing the following
- Merge all invalid files for **script v0.0.1** into one file ``invalid_script_v0.0.1.yml`` since we can process multiple tests 
- Add test for valid files for **script v0.0.1** called ``valid_script_v0.0.1.yml``
- Add method ``check_valid_recipes``  and ``check_invalid_recipes`` that iterates over all invalid and valid recipes.
- Refactor ``test_organization.py`` to do the following
  - Move field check  ``"[$id", "$schema", "title", "propertyNames", "properties"]`` at top since they are top of schema makes for top-down assertion easier for us to understand as we validate schema.
  - Utilize for loop  over ``found = ["type","description","maintainers","env","executor","shell","pre_run", "run","post_run","status",]`` to check all keys instead of doing them one assert at time such as ``assert env in loaded['properties']``
  - At most do a check for key and type validation check for all keys. I didn't do the nested keys in ``status`` since its not done.
  - There was no assert for validating ``executor`` key. 
  - There was no assert for for ``type`` pattern check which is done via ``assert loaded["properties"]["type"]["pattern"] == "^script$"``
  - Perform a for loop over like ``type`` objects to do more efficient assert such as 
     ```
            for section in ["type", "description", "pre_run", "post_run", "shell", "executor"]:
                assert loaded["properties"][section]["type"] == "string"
     ```

Now the issue is when running test i am not getting an exception raised when testing ``invalid_regex_stream``. 

```
            for name in content.keys():
                with pytest.raises(ValidationError) as excinfo:
>                   validate(instance=content[name], schema=loaded)
E                   Failed: DID NOT RAISE <class 'jsonschema.exceptions.ValidationError'>

.github/tests/test_script.py:52: Failed
------------------------------------- Captured stdout call --------------------------------------
Root of testing is /mxg-hpc/users/ssi29/schemas
Testing schema script
Getting version of script-v0.0.1.schema.json
Checking that invalids exist for script-v0.0.1.schema.json
Testing multiline_run from recipe valid_script_v0.0.1.yml should be valid
Testing single_command_run from recipe valid_script_v0.0.1.yml should be valid
Testing declare_env from recipe valid_script_v0.0.1.yml should be valid
Testing declare_shell from recipe valid_script_v0.0.1.yml should be valid
Testing declare_maintainer from recipe valid_script_v0.0.1.yml should be valid
Testing status_returncode from recipe valid_script_v0.0.1.yml should be valid
Testing status_regex from recipe valid_script_v0.0.1.yml should be valid
Testing invalid_test_name_&!@#$% from recipe invalid_script_v0.0.1.yml should be invalid
Testing invalid_bash from recipe invalid_script_v0.0.1.yml should be invalid
Testing missing_run_key from recipe invalid_script_v0.0.1.yml should be invalid
Testing invalid_maintainer from recipe invalid_script_v0.0.1.yml should be invalid
Testing invalid_env_type from recipe invalid_script_v0.0.1.yml should be invalid
Testing invalid_description from recipe invalid_script_v0.0.1.yml should be invalid
```

The test below should fail because ``stream`` has enum on ``stdout``, ``stderr`` so ``file`` is invalid. 

```
invalid_regex_stream:
  type: script
  description: This test fails because of invalid regex stream
  run: hostname
  status:
    regex:
      stream: file
      exp: "world$"
```

I also find that any of the subsequent test afterwards are not raising exception, it as if ``anyOf`` is  valid against against one of the two sub-schema but dont know why.  regex is an object while returncode is a integer so they are not like objects for status to make a check.

What i want is ``returncode`` and ``regex`` can't be specified together hence their is a mutual exclusion test 

```

mutual_exclusion_status:
  type: script
  description: This test fails because returncode and regex specified
  run: hostname
  status:
    returncode: 1
    regex:
      stream: stdout
      exp: "^hello"
```

Let me know you have any comments on any other changes unrelated to ``status`` that needs to be fixed. If there is way to get all valid and invalid checks passed i think we would be confident the schema is working as expected.

### CODE REVIEW

**Comment by vsoch** [comment]
Seems okay so far. What specifically do you want feedback on?

**Comment by vsoch** [comment]
I don't think you want anyOf for the status object, just put the fields directly below it, and neither are required.

**Comment by shahzebsiddiqui** [comment]
> Seems okay so far. What specifically do you want feedback on?

Ok great, if my reasoning makes sense in the PR description then we are good.  There were some changes to ``test_organization.py`` and ``test_script.py`` that were unrelated to the ``status`` key which were changes i tacked on to this PR.

**Comment by shahzebsiddiqui** [comment]
> I don't think you want anyOf for the status object, just put the fields directly below it, and neither are required.

Okay so i removed ``anyOf`` now all the test work except for mutual exclusion one, see https://github.com/shahzebsiddiqui/schemas/blob/4f32a68c07b0f93868feefc81164ff980cbd4406/.github/tests/invalid/script/0.0.1/invalid_script_v0.0.1.yml#L63

I guess question is do we want schema to complain about mutual exclusion between ``returncode`` and ``regex`` or is this acceptable. If this is acceptable then buildtest needs to pick which one to use for checking test status.

**Comment by vsoch** [comment]
> I guess question is do we want schema to complain about mutual exclusion between returncode and regex or is this acceptable. If this is acceptable then buildtest needs to pick which one to use for checking test status.

I think it's reasonable that a user would want to check BOTH a return code and the stderr /stdout. I would say that if both are defined, both must pass to get status "PASSED" and if either fails, the test should not be considered successful. And if course if only one is defined, just use that.

**Comment by shahzebsiddiqui** [comment]
> > I guess question is do we want schema to complain about mutual exclusion between returncode and regex or is this acceptable. If this is acceptable then buildtest needs to pick which one to use for checking test status.
> 
> I think it's reasonable that a user would want to check BOTH a return code and the stderr /stdout. I would say that if both are defined, both must pass to get status "PASSED" and if either fails, the test should not be considered successful. And if course if only one is defined, just use that.

Makes sense to me, in that case i will the one failed test to a valid test. Initially i expected mutual exclusion between returncode and regex but based on what i understand we can allow for both status checks before reporting test is passed.

**Comment by vsoch** [comment]
I would put the status check for passed() and failed() as a function of the builder, and the result depends on the buildspec (what is allowed) and exposes the same interface (passed or failed) to the client. E.g.,:
```
builder.passed()
True

or builder.failed()
False
````
And then say, for the script type if it allows returncode and a regular expression, you put the logic for checking both inside that function. E.g.,

```
# case 1: both are defined, both must pass to get "passed" is True
# case 2: one is defined, just that one must pass.

And then the base class would just use the return code. This is what I think would be the most logical implementation, because undoubtable different build types will have different criteria for passing or failing, but they must be interacted with using the same interface.
```

**Comment by shahzebsiddiqui** [comment]
This exercise was quite interesting, i am now getting a hang of these regtest. I can see this going to be very time consuming trying to check every single key and type. Considering that the regression test is tied specific json schema that at a particular point in time i can see these ``assert`` statement to be constantly changed. I anticipate the json schema to go through several revisions in the beginning. 

The schema would be changed to address a new capability in a Buildspec and that would be irrespective of the regression test, since there is no source code. I see these regtest to ensure confidence of each schema.

I like the design what i tried to do where we can write a list of valid and invalid tests and let json schema check for all errors. There are going to be lot's of edge cases we would need to think for every key and that can get challenging to keep track of. 

Anyhow, i now see the point of framework separated from schema.

File: .github/tests/test_organization.py [hunk 1]
```diff
@@ -75,42 +75,19 @@ def test_schema_naming(tmp_path):
             # Assert it loads with jsonschema
             schema_file = os.path.join(schema_dir, schema)
             loaded = load_schema(schema_file)
-
+            # print(json.dumps(loaded, indent=2))
```
Comment (vsoch) [inline .github/tests/test_organization.py:78]: should this be commented out (or just deleted, better?)

File: .github/tests/test_organization.py [hunk 2]
```diff
@@ -123,6 +100,71 @@ def test_schema_naming(tmp_path):
             assert match.groups()
             version = match["version"]
 
+            assert "required" in loaded
+            # check "type" and "run" are required keys
+            assert "type" in loaded["required"]
+            assert "run" in loaded["required"]
+
+            # check all keys are properties
```
Comment (vsoch) [inline .github/tests/test_organization.py:108]: Just to clarify - we are on the same page that these are required (or defined to be optional) currently for all schemas.

File: .github/tests/test_script.py [hunk 3]
```diff
@@ -35,6 +35,41 @@ def load_recipe(path):
     return content
 
 
+def check_invalid_recipes(recipes, invalids, loaded, version):
```
Comment (vsoch) [inline .github/tests/test_script.py:38]: And then the test_script is just for the script type, and we can follow this test_<name> convention for upcoming schemas.

**Comment by vsoch** [comment]
> Anyhow, i now see the point of framework separated from schema.

I can't tell you how happy I am that you can see this now! Imagine that the project gets much larger, each schema development is important, a lot of detailed work, and requires substantial testing, and it would make buildtest a huge hairball / much harder to develop. On the other hand, with this setup, all the work and testing is done here, and when a merge is done for a "final" version, then there is one simple, clean PR to add it to buildtest (with an executor class) and there is assurance that the schema files, at least per the fields required, etc., are correct. 

The way we are developing now is still very early, so we aren't as 1:1 when it comes to developing a schema, then adding it to buildtest. But it will get closer to that, and (over time) this repository won't have as much activity or change, because the yaml specs won't change.

File: .github/tests/test_organization.py (hunk 1 above)
Comment (shahzebsiddiqui) [reply .github/tests/test_organization.py:78]: my bad, that was me troubleshooting. I got this fixed now

File: .github/tests/test_organization.py (hunk 2 above)
Comment (shahzebsiddiqui) [reply .github/tests/test_organization.py:108]: oh that list was meant for script schema, but i see now that for loop is going over entire ``schema`` directory. Since there is only one schema this works, but it won't going forward with other schema files. I'd prefer to refactor this later because i would like to work with notion of writing a regtest per schema in a separate file like ``test_script.py``. Right now it was not clear which keys are global and which are not and i noticed when i started there were several key checks in ``test_organization.py`` so i figured i add all the checks in the same file. 

Maybe all the global keys shared by all schema should be tested in ``test_global.py`` and ``test_organization`` just checks the directory layout and names check (schema-v0.0.1.schema.json``).

I'd want to refactor these later in another PR.

File: .github/tests/test_script.py (hunk 3 above)
Comment (shahzebsiddiqui) [reply .github/tests/test_script.py:38]: exactly, every schema should have atleast one file valid and invalid Buildspec with list of examples to test all the valid and invalid cases. The method ``check_invalid_recipes`` and ``check_valid_recipes`` can work for any version provided the inputs are passed to this method. Hope this design makes sense.

File: .github/tests/test_organization.py (hunk 2 above)
Comment (vsoch) [reply .github/tests/test_organization.py:108]: okay great - this sounds like a good plan, and what I wanted to talk about. We didn't have the concept of the global schema developed when I created test_organization.py, so I also agree about refactoring that file. But out of scope for this PR.

File: .github/tests/test_organization.py (hunk 1 above)
Comment (vsoch) [reply .github/tests/test_organization.py:78]: Was this removed? I approved, but conditional on removing this line.

Comment (shahzebsiddiqui) [reply .github/tests/test_organization.py]: This is fixed now

**Comment by vsoch** [comment]
Good for merge _after tests pass again_.

**Comment by shahzebsiddiqui** [comment]
> > Anyhow, i now see the point of framework separated from schema.
> 
> I can't tell you how happy I am that you can see this now! Imagine that the project gets much larger, each schema development is important, a lot of detailed work, and requires substantial testing, and it would make buildtest a huge hairball / much harder to develop. On the other hand, with this setup, all the work and testing is done here, and when a merge is done for a "final" version, then there is one simple, clean PR to add it to buildtest (with an executor class) and there is assurance that the schema files, at least per the fields required, etc., are correct.
> 
> The way we are developing now is still very early, so we aren't as 1:1 when it comes to developing a schema, then adding it to buildtest. But it will get closer to that, and (over time) this repository won't have as much activity or change, because the yaml specs won't change.

Yea i can see value of this repo down the road when framework. I can see a small change in schema can result in unintended consequences on the framework so having this separate is good idea.

 
I can see myself keeping more than one version of a schema ``script v0.0.1`` while working on a significant change to script version ``0.0.2`` without breaking the framework. Though i don't ever see many versions of a schema, as you said schema won't change it's not software. For instance kubernettes api versions https://kubernetes.io/docs/concepts/overview/kubernetes-api/  are simple as ``api/v1``. I think we can do better with the ``version`` key. I guess we first need more schema types for different things before we make this change.
